File: app/src/main/java/me/neocode/slftool/MainActivity.java
package me.neocode.slftool;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.content.pm.ActivityInfo;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.URL;
import java.util.Random;

public class MainActivity extends AppCompatActivity {

    JSONObject jsonData;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        getSupportActionBar().hide();
        setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_LOCKED);

        final Button regenerate = findViewById(R.id.regenerate);
        final EditText letter = findViewById(R.id.letter);
        final TextView view = findViewById(R.id.textView);
        final ProgressBar progressBar = findViewById(R.id.progressBar);

        new Thread(new Runnable() {
            @Override
            public void run() {
                try {
                    URL link = new URL("https://slftool.github.io/data.json");
                    BufferedReader in = new BufferedReader(new InputStreamReader(link.openStream()));
                    String line;
                    String all = "";

                    while ((line = in.readLine()) != null) {
                        all = all + line;
                    }
                    jsonData = new JSONObject(all);
                    runOnUiThread(new Runnable() {
                        public void run() {
                            regenerate.setEnabled(true);
                            progressBar.setVisibility(View.INVISIBLE);
                            regenerate.setVisibility(View.VISIBLE);
                            letter.setVisibility(View.VISIBLE);
                            Toast.makeText(getApplicationContext(), "Datenbank erfolgreich geladen!", Toast.LENGTH_SHORT).show();
                        }
                    });
                } catch (IOException | JSONException e) {
                    e.printStackTrace();
                    runOnUiThread(new Runnable() {
                        public void run() {
                            Toast.makeText(getApplicationContext(), "Beim laden der Datenbank ist ein Fehler aufgetreten! kein Internet?", Toast.LENGTH_LONG).show();
                        }
                    });
                }
            }
        }).start();

        regenerate.setOnClickListener(new Button.OnClickListener() {
            @SuppressLint("SetTextI18n")
            public void onClick(View v) {
                String letterName = letter.getText().toString();
                letterName = letterName.toLowerCase();
                if ("abcdefghijklmnopqrstuvwxyz".contains(letterName)) {
                    try {
                        hideKeyboard();

                        JSONArray stadt = jsonData.getJSONObject(letterName).getJSONArray("stadt");
                        JSONArray land = jsonData.getJSONObject(letterName).getJSONArray("land");
                        JSONArray fluss = jsonData.getJSONObject(letterName).getJSONArray("fluss");
                        JSONArray name = jsonData.getJSONObject(letterName).getJSONArray("name");
                        JSONArray beruf = jsonData.getJSONObject(letterName).getJSONArray("beruf");
                        JSONArray tier = jsonData.getJSONObject(letterName).getJSONArray("tier");
                        JSONArray marke = jsonData.getJSONObject(letterName).getJSONArray("marke");
                        JSONArray pflanze = jsonData.getJSONObject(letterName).getJSONArray("pflanze");

                        view.setText("Stadt: " + stadt.getString(getRandomNumberInRange(0, stadt.length())) +
                                "\nLand: " + land.getString(getRandomNumberInRange(0, land.length())) +
                                "\nFluss: " + fluss.getString(getRandomNumberInRange(0, fluss.length())) +
                                "\nName: " + name.getString(getRandomNumberInRange(0, name.length())) +
                                "\nBeruf: " + beruf.getString(getRandomNumberInRange(0, beruf.length())) +
                                "\nTier: " + tier.getString(getRandomNumberInRange(0, tier.length()))+
                                "\nMarke: " + marke.getString(getRandomNumberInRange(0, marke.length()))+
                                "\nPflanze: " + pflanze.getString(getRandomNumberInRange(0, pflanze.length())));
                    } catch (JSONException e1) {
                        e1.printStackTrace();
                    }
                }
            }
        });
    }

    private void hideKeyboard() {
        View view = findViewById(android.R.id.content);
        if (view != null) {
            InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
            imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
        }
    }

    private int getRandomNumberInRange(int min, int max) {
        Random r = new Random();
        return r.nextInt((max - min)) + min;
    }
}

File: app/src/main/java/me/neocode/slftool/MainActivity.kt
package me.neocode.slftool

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.WindowCompat
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import me.neocode.slftool.databinding.ActivityMainBinding
import okhttp3.*
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.io.IOException
import kotlin.system.exitProcess

class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding
    private lateinit var adapter: Adapter
    private lateinit var json: JSONObject
    private val icons: HashMap<String, Int> = HashMap()

    override fun onCreate(savedInstanceState: Bundle?) {
        WindowCompat.setDecorFitsSystemWindows(window, false)
        super.onCreate(savedInstanceState)

        icons["stadt"] = R.drawable.ic_baseline_location_city_24
        icons["land"] = R.drawable.ic_baseline_flag_24
        icons["fluss"] = R.drawable.ic_baseline_waves_24
        icons["name"] = R.drawable.ic_baseline_person_24
        icons["beruf"] = R.drawable.ic_baseline_work_24
        icons["tier"] = R.drawable.ic_baseline_pets_24
        icons["marke"] = R.drawable.ic_baseline_store_24
        icons["pflanze"] = R.drawable.ic_baseline_grass_24

        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        binding.fab.setOnClickListener {
            changeWords()
        }

        binding.refreshLayout.setOnRefreshListener {
            changeWords()
            binding.refreshLayout.isRefreshing = false
        }

        binding.letter.addTextChangedListener(object : TextWatcher {
            var changed = false
            override fun beforeTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {}
            override fun onTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {}
            override fun afterTextChanged(e: Editable?) {
                if(!changed && (binding.letter.text?.length ?: 0) > 0) {
                    changed = true
                    binding.letter.setText(binding.letter.text.toString().takeLast(1))
                    binding.letter.setSelection(1)
                    changeWords()
                }else{
                    changed = false
                }
            }

        })

        binding.recycler.layoutManager = LinearLayoutManager(this)

        val data = ArrayList<ItemModel>()
        data.add(ItemModel(R.drawable.ic_baseline_help_24, getString(R.string.letter_required), ""))
        adapter = Adapter(data)
        binding.recycler.adapter = adapter

        val client: OkHttpClient = OkHttpClient.Builder()
            .cache(
                Cache(
                    directory = File(application.cacheDir, "http_cache"),
                    // $0.05 worth of phone storage in 2020
                    maxSize = 50L * 1024L * 1024L // 50 MiB
                )
            )
            .build()

        val request = Request.Builder()
            .url("https://slftool.github.io/data.json")
            .build()

        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                e.printStackTrace()
                runOnUiThread {
                    MaterialAlertDialogBuilder(this@MainActivity)
                        .setTitle(getString(R.string.no_internet))
                        .setMessage(getString(R.string.loading_error))
                        .setNeutralButton(getString(R.string.ok)) { _, _ ->
                            finishAffinity()
                            exitProcess(0)
                        }
                        .show()
                }
            }

            override fun onResponse(call: Call, response: Response) {
                response.use {
                    if (!response.isSuccessful) throw IOException("Unexpected code $response")

                    json = JSONObject(response.body!!.string())
                    runOnUiThread {
                        Toast.makeText(applicationContext, R.string.loaded_successfully, Toast.LENGTH_SHORT).show()
                    }
                }
            }
        })
    }

    private fun changeWords() {
        val letter: String = binding.letter.text.toString().lowercase()
        val data = ArrayList<ItemModel>()

        if (this::json.isInitialized && json.has(letter)) {
            val letterJSON: JSONObject = json.getJSONObject(letter)

            letterJSON.keys().forEach { key ->
                val obj: JSONArray = letterJSON.getJSONArray(key)
                val random: String = obj.getString((0 until obj.length()).random())

                var icon: Int = if(icons.containsKey(key)) {
                    icons[key]!!
                }else{
                    R.drawable.ic_baseline_help_24
                }

                data.add(ItemModel(icon, random, key.capitalize()))
            }
        }else{
            data.add(ItemModel(R.drawable.ic_baseline_help_24, getString(R.string.letter_required), ""))
        }

        adapter.setItems(data)
    }

}